### PR TITLE
Access Point files for the backend

### PR DESCRIPTION
**Pull request by anmolkabra**
Tested, working implementation of Access Point setup.
### Dependencies
* hostapd
* dnsmasq
* apache2
Note: These must be installed before proceeding. They don't come pre-installed with the pi, so one must connect to the internet and install them. This is ironic to our motive, but I will work on pre-installing these packages in a OS image and shipping that OS image.

### How it works (with the current setup)
On a raspberry pi, navigate to `cs-minibot/minibot/setup` and run `sudo ./handle_startup_wifi_connection.sh` to setup a Access Point. If the device is connected to the internet already, then the Access Point setup will be skipped (why does one need to make pi an access point if it's already accessible through the internet?). The latter will change as users might want their pi's to connect to another wifi network.
* If the pi couldn't connect to the internet in 3 seconds, then the Wifi Access Point will be setup. It may throw some warnings related to the hostapd and dnsmasq not being native processes, but those can be ignored.
* `sudo reboot` the pi and you'll be able to connect to the network named 'minibotAP' with CornellCup-Web's password. Not writing here for obvious security issues.

Currently, the Web server is not actively asking the user for Wifi credentials. This is a basic construct of setting and cleaning up the Access Point settings.
To clean the Access Point settings, run `sudo ./handle_startup_wifi_connection.sh clean`. 

The `sudo` is important in running those files.

### CODE REVIEW

**Comment by ScottSWu** [comment]
For future, we can replace apache web server with python+tornado

**Comment by chynu** [comment]
Guh how does one test this on the bot if it has already been connected to the internet to download the repo that has the code on it lol

**Comment by anmolkabra** [comment]
Yeah that's an issue. The access point will only be created if the pi is not connected to the internet, as of current scripting. Down load the repo, checkout the branch, disconnect from the internet and run the script. Might have to connect to a monitor (ssh won't work) as when you disconnect from the internet, you'll lose control of the pi. I'll add some scripts later to automatically run this script Everytime the pi boots up

**Comment by ScottSWu** [comment]
From stackoverflow https://stackoverflow.com/a/26820300

Add something like this to the startup script:
```
wget -q --spider https://github.com

if [ $? -eq 0 ]; then
    # Clone repo, etc.
else
    # Start access point
fi
```

**Comment by anmolkabra** [comment]
No. 

1. If wget couldn't get the repo in the first command, the first branch will be taken. Of it couldn't connect to the internet in the first command, how do you expect the script to clone the repo? Correct me if I'm interpreting it incorrectly.

2. The access point setup is very clumsy right now as things are not automatic. I would love to add these scripts only when the website integration is done. Celine and I will work on python+tornado, as you suggested.

I can talk about this more.

**Comment by ScottSWu** [comment]
Oh the wget command is just to check if there is an internet connection. You can replace github.com with any other url, but it'd make sense that github itself is up.

The if statement checks the status code if the wget function. Typically the a status code of 0 means "success".

**Comment by anmolkabra** [comment]
Yes yes I understand that and I think you're right. I think I messed up what I meant. Ignore the previous comment about this, and consider this:

Since the access point setup comes with the repo, we have two problems here:
1. If the pi is able to connect to the internet, then it clones the repo, fine.
2. If the pi is not able to connect to the internet, then logic flows to the second branch. But the AP won't work since the files and scripts are not yet present! (They come with the repo which is non-existent on a fresh boot).

In other words, this init script will only work when the repo is already present on the pi. We have a much bigger problem when the user fresh boots the pi - no internet, no repo, and the pi is dirt for time being. For this AP to run out of box, the pi must come with our software installed - a custom OS! This is nother issue though.

**I will add that init script when we have something similar to what the end product will look like.**

**Comment by anmolkabra** [comment]
Additionally, I have similar logic (check wifi, do something if connected etc.) in `handle_startup_wifi_connection.sh`. I will add that logic to an init script afterwards.

**Comment by chynu** [comment]
Perhaps we should review and merge this and make sure this works with the rest of the code before moving forth and implementing a tornado version, since we have an upcoming deadline (Monday).

**Comment by chynu** [comment]
Yeah there's a bit of a circular problem here. There's no need to have any of this if it's already connected to the internet, and the script is tied to the repo which you can only get from the internet lol

We'll probably have to ship SD cards flashed _with_ the repo on it, alongside everything else (or like an OS like anmol said). A lot of our testing is sort of "developer mode" anyway

How feasible is making our own OS? lol

**Comment by anmolkabra** [comment]
Okay so this is done now. Using a tornado web app at 10.0.0.1:8080 instead of Apache server. I have read that `/etc/rc.local` is executed last in boot, so I use that file to automate the wifi and access points checks. This is different from using a script in `/etc/init.d/` or something.

Please review and let me know if something's wrong. Added a readme to the folder so people can go to the correct file and debug.

**Comment by anmolkabra** [comment]
Addresses Issue https://github.com/cornell-cup/cs-minibot/issues/57

**Review by chynu — APPROVED** [review]
cool
